File: maximum-performance-of-a-team/maximum-performance-of-a-team.java
class Solution {
    public int maxPerformance(int n, int[] speed, int[] efficiency, int k) {
        
        Engineer[] eng = new Engineer[n];
		for (int i = 0; i < n; i++) {
			eng[i] = new Engineer(speed[i], efficiency[i]);
		}
		Arrays.sort(eng);
		PriorityQueue<Engineer> pq = new PriorityQueue<Engineer>((a, b) -> a.speed - b.speed);

	    long max = 0;
		long sum = 0;

		for (int i = 0; i < n; i++) {
			pq.add(eng[i]);
			sum += eng[i].speed;
			if (pq.size() > k) {
				sum -= pq.poll().speed;
			}
			
			long tempPerformace = sum * (long) eng[i].eff;
			max = Math.max(max,tempPerformace);

		}

		 return (int) (max % 1000000007);
        
    }
    
    static class Engineer implements Comparable<Engineer> {
		int speed;
		int eff;

		Engineer(int speed, int eff) {
			this.speed = speed;
			this.eff = eff;
		}

		public int compareTo(Engineer e) {
			return e.eff - this.eff;

		}
	}
}

File: 1499-maximum-performance-of-a-team/maximum-performance-of-a-team.cpp
class Solution {
public:
    typedef long long ll;
    int maxPerformance(int n, vector<int>& speed, vector<int>& eff, int k) {

        ll ans = 0, sum=0;
        int M=1e9+7;
        vector<pair<int,int>>v; // stores (eff, speed) in sorted order

        //stores max k elements with eff greater than curr index and max speeds that are possible
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>>pq;
        
        for(int i=0;i<n;i++){
            v.push_back({eff[i], speed[i]});
        }
        sort(v.begin(),v.end());

        for(int i=n-1;i>=0;i--){
            if(pq.size()>=k){
                sum -= pq.top().first;
                pq.pop();
            }
            sum += v[i].second;
            pq.push({v[i].second,v[i].first});
            ans = max(ans,sum*v[i].first);
        }
        return ans%M;
    }
};